### ISSUE TITLE
Unwanted exclusion blocks in generated POM

### ISSUE THREAD

**Issue by aalmiray** [gradle/gradle#12909]
### Expected Behavior
Configuring the `maven-publish` plugin and adding exclusions via `configurations.all` results in exclusions being applied to all `<dependency>` elements, not the matching ones.

Take the following sample build file

    plugins {
        id 'java-library'
        id 'maven-publish'
    }
    
    group   = 'com.acme'
    version = '0.0.0-SNAPSHOT'
    
    repositories {
        jcenter()
    }
    
    dependencies {
        api 'commons-lang:commons-lang:2.6'
        api 'commons-beanutils:commons-beanutils:1.9.4'
        api 'junit:junit:4.13'
    }
    
    publishing {
        publications {
            maven(MavenPublication) {
                from components.java
            }
        }
    }
    
    configurations.all {
        exclude group: 'commons-logging', module: 'commons-logging'
    }

commons-logging is not desired as part of the dependency graph. Without the exclusions the resolved compileClasspath looks like this

    compileClasspath - Compile classpath for source set 'main'.
    +--- commons-lang:commons-lang:2.6
    +--- commons-beanutils:commons-beanutils:1.9.4
    |    +--- commons-logging:commons-logging:1.2
    |    \--- commons-collections:commons-collections:3.2.2
    \--- junit:junit:4.13
         \--- org.hamcrest:hamcrest-core:1.3

Once the exclusion is applied and `gradle generatePomFileForMavenPublication` is invoked the expectation would be to see a POM file like the following one

    <?xml version="1.0" encoding="UTF-8"?>
    <project xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd" xmlns="http://maven.apache.org/POM/4.0.0"
        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
      
      
      
      
      
      <modelVersion>4.0.0</modelVersion>
      <groupId>com.acme</groupId>
      <artifactId>foo</artifactId>
      <version>0.0.0-SNAPSHOT</version>
      <dependencies>
        <dependency>
          <groupId>commons-lang</groupId>
          <artifactId>commons-lang</artifactId>
          <version>2.6</version>
          <scope>compile</scope>
        </dependency>
        <dependency>
          <groupId>commons-beanutils</groupId>
          <artifactId>commons-beanutils</artifactId>
          <version>1.9.4</version>
          <scope>compile</scope>
          <exclusions>
            <exclusion>
              <artifactId>commons-logging</artifactId>
              <groupId>commons-logging</groupId>
            </exclusion>
          </exclusions>
        </dependency>
        <dependency>
          <groupId>junit</groupId>
          <artifactId>junit</artifactId>
          <version>4.13</version>
          <scope>compile</scope>
        </dependency>
      </dependencies>
    </project>

### Current Behavior
Every `<dependency>` element in the generated POM has an exclusion, even if it's not found in its transitive dependencies

    <?xml version="1.0" encoding="UTF-8"?>
    <project xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd" xmlns="http://maven.apache.org/POM/4.0.0"
        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
      
      
      
      
      
      <modelVersion>4.0.0</modelVersion>
      <groupId>com.acme</groupId>
      <artifactId>foo</artifactId>
      <version>0.0.0-SNAPSHOT</version>
      <dependencies>
        <dependency>
          <groupId>commons-lang</groupId>
          <artifactId>commons-lang</artifactId>
          <version>2.6</version>
          <scope>compile</scope>
          <exclusions>
            <exclusion>
              <artifactId>commons-logging</artifactId>
              <groupId>commons-logging</groupId>
            </exclusion>
          </exclusions>
        </dependency>
        <dependency>
          <groupId>commons-beanutils</groupId>
          <artifactId>commons-beanutils</artifactId>
          <version>1.9.4</version>
          <scope>compile</scope>
          <exclusions>
            <exclusion>
              <artifactId>commons-logging</artifactId>
              <groupId>commons-logging</groupId>
            </exclusion>
          </exclusions>
        </dependency>
        <dependency>
          <groupId>junit</groupId>
          <artifactId>junit</artifactId>
          <version>4.13</version>
          <scope>compile</scope>
          <exclusions>
            <exclusion>
              <artifactId>commons-logging</artifactId>
              <groupId>commons-logging</groupId>
            </exclusion>
          </exclusions>
        </dependency>
      </dependencies>
    </project>

### Context
Adding multiple exclusions in a medium to big size project results in bigger POM files. That's a wast of space and transferred bytes, not to mention tools that have to parse and process exclusions that are irrelevant.

### Steps to Reproduce 
The provided build file above is enough to reproduce this issue.

### Your Environment

Build scan URL: https://scans.gradle.com/s/2xjajhrlqd6ds

**Comment by ljacomet** [comment]
While this could indeed be an improvement, there is a possible workaround: use dependency level excludes.

So that the Gradle build itself will be faster because it will not have to carry the configuration level excludes on all dependency paths, including those that are not needed.
And the Maven POM will then have the excludes at the right place as well.

Unless you are willing to attempt a fix for this solution, I don't see this changing any time soon.

**Comment by melix** [comment]
It works as expected. When you do:

`configurations.all { exclude .... }`

you're asking Gradle to exclude a dependency from _all_ resolved configurations in _all_ paths. Gradle cannot know in advance if a path _will_ trigger the inclusion of a dependency or not, because it depends on the other dependencies present in the graph (b/c of the game of conflict resolution).

Said differently, we can't "know" that an exclusion will apply to a single node or not, even if in theory, it should be possible. However, what I'm saying is that even if we _knew_, it would be inconsistent, semantically wise, to exclude only for a single dependency, because that's not the semantics of the original exclusion.

**Comment by aalmiray** [comment]
@ljacomet that's right, there is a workaround using explicit exclusions per dependency. Which is OK when you have a few dependencies in the project(s), but a nightmare when you have many, hundreds; which is why a global excusion mechanism such as `configurations.all { }` is desirable.

@melix I don't expect Gradle to change how it performs dependency resolution. Gradle can create and handle the dependency graph in the best way it needs to, no need to eagerly optimize the graph for what a potential consumer (during the build) would or could do. However concerning the interface with other tools (in this case the generated POM which will be consumed by tools other than Gradle).

What I'm asking is to update the mechanism when the POM is generated by the `maven-publish` plugin, as I suspect this is done deep in the plugin. I'm not asking to change the rules of dependency resolution. Does this make sense?

**Comment by melix** [comment]
>  I'm not asking to change the rules of dependency resolution. Does this make sense?

It does but what I'm saying is that if you don't publish "global" exclusions defined like this on _all_ dependencies, then you're breaking the symmetry between publication and consumption. Basically it means that consumers wouldn't get the same results as you do on the producer side, which isn't nice. It's _always_ the case in practice, but for exclusions, which are a bit of a pain tbh, it's even worse. If you resolve from the producer side, the "global" exclusions are excluded from _all_ edges. If you, instead (and again it assumes that we would be able to figure out the edges affected by an exclude at publication time, which involves resolving and analyzing the graph), publish excludes only on specific dependencies, then the exclusions are _not_ global anymore. Which changes the semantics for the consumer. If the goal is to "clean up" the generated POM, I think it's not a good idea. If the goal is to have something "closer to the reality", then it's a misconception to think it is, because it effectively depends on the whole graph.

**Comment by aalmiray** [comment]
Bingo! Exclusions need to be published to consumers, however the current algorithm casts a wide net. Forget me if I misunderstood but isn't the case that by the time dependencies are added to a generated POM, the participating configurations have been resolved already?

Gradle knows what several things at this point:
- it _knows_ the dependencies that were defined in the build (these are the ones found in the build script before resolution).
- it _knows_ the resolved artifacts per configuration, this includes direct and transitive dependencies.
- it's aware of the global exclusions.

This information could be used at the time of POM generation to determine if global exclusions match the subgraph per direct dependency. Both `dependencies` and `dependencyInsight` render portions of the graph as needed; it follows that similar code could be put in place during POM generation. Perhaps I'm missing something here as I'm no expert in Gradle dependency resolution mechanics, but from an eagle's eye perspective I'd say this is feasible, isn't it? Albeit I've no clue how much effort it would require to make it happen.

**Comment by melix** [comment]
> Bingo! Exclusions need to be published to consumers, however the current algorithm casts a wide net.

It publishes exactly what you asked for, consistently.

> Forget me if I misunderstood but isn't the case that by the time dependencies are added to a generated POM, the participating configurations have been resolved already?

Absolutely not. You can publish a library without resolving anything. What is published corresponds to the _declared_ versions (it has always been the case, even with the `maven` plugin, and Maven does the same). Recently we added the ability to [publish resolved versions](https://docs.gradle.org/current/userguide/publishing_maven.html#publishing_maven:resolved_dependencies) but it's a "guess" what to resolve to get meaningful results. The defaults in Gradle are to use `compileClasspath` and `runtimeClasspath`, but it could be something else.

There are lots of edge cases with resolved versions (think about dependency substitutions), so it's not as simple as it seems.

> This information could be used at the time of POM generation to determine if global exclusions match the subgraph per direct dependency. 

Yes and no, because the "shape" of the graph changes as resolution goes (as we discover transitive dependencies and resolve conflicts). It's not uncommon (and actually fairly frequent) that excludes are added or removed _during resolution_. And not mentioning that adding excludes on the configuration level actually.... changes the graph resolution itself!

And again, I don't think this is something we should change. You didn't really explain _why_ you want this: what we do is _consistent_ on the producer and consumer sides. If you put a configuration level exclude, it's global, independently of the path, so for the consumer to have the same result (at least have the most guarantees it can, this is not the reality), we _have_ to put it on every dependency.

**Comment by aalmiray** [comment]
I see. As I said I’m no Gradle dependency resolution expert. As @ljacomet commented, local exclusions are a workaround, and that’s OK for a small dependency graph like the example shown above. But what if I **need** to exclude the same dependency from many places in a huge graph? 

Say you have to exclude `commons-logging:commons-logging` no matter where it may come from. The current options are:

1. Search for all places where that dependency is coming from, probably using `dependencyInsight` and `dependencies` (a couple of times to get all configurations). This is likely an arduous task depending on the size of the graph but nets you the exact exclusions you need and want in the generated POM. Of course this only works as long as no addition deps are added to the project(s), in which case the search and exclusion work must be performed again, and again. 

2. Apply a global exclude. This nets you the required exclusions today and future ones, no matter how the dependency graph changes. The downside is unneeded/unwanted exclusions in the generated POM. 

Are there any other options I missed?

**Comment by melix** [comment]
Excludes should be last resort. For `commons-logging` you probably want to use something like @ljacomet 's [logging plugin](https://blog.gradle.org/addressing-logging-complexity-capabilities) instead, which avoids exclusions and performs proper dependency resolution.

Otherwise options should be listed at https://docs.gradle.org/current/userguide/dependency_downgrade_and_exclude.html

**Comment by aalmiray** [comment]
The use of `commons-logging` is an example, it could be any dependency. If I want to exclude Guava from anywhere in the graph, does this mean a custom plugin like the `logging plugin` need to be built? I don't think so, that would lead to an explosion of plugins excluding a very specific set of dependencies, then someone will realize that's silly and create a more generic plugin, then realize that's even sillier.

**Comment by melix** [comment]
> does this mean a custom plugin like the logging plugin need to be built but only for Guava?

No, that's not the point. The point is _why_ do you need to exclude a dependency. Because depending on the _why_ there's a better solution than another. The plugin I mention is an example where better modelling improves the situation for everyone, because instead of everybody excluding their loggers and re-introducing because nobody agrees, the solution here is _not_ excludes because that's not what you want. The solution is to model the fact that there are many different loggers and that a particular consumer _needs to choose one_. The plugin does that, and if the producers (the logger framework authors) agree on the model, then the plugin becomes redundant. Our goal is to move beyond the silly exclude model which doesn't tell _why_ you need to exclude so the engine cannot take smart decisions and everybody is forced to exclude/reintroduce/patch metadata.

If you want to exclude Guava, then maybe there's another reason. Maybe the code path you exercise don't need any of the Guava code which is brought transitively, and then it's a _good_ use case for an `exclude`, but on the dependency level. Maybe you want to exclude because the dependency shouldn't have been there in the beginning, it's universally true independently of the consumer, it's bad metadata in which case the solution is a component metadata rule (fix the pom), etc...

In short, instead of a hammer, we provide the tools to fix each problem. And if one problem is too hard/too complex to fix with the tools we provide, we need to improve them.

**Comment by aalmiray** [comment]
I guess what I'm saying, until the time comes (if it comes) for the metadata format to be adopted by the tooling ecosystem, there has to be a migration period for the time being. A period of compatibility bridges, warnings, deprecations, and so forth.

Given that Gradle has proved to have more speed in this area when compared to other tools, I'm asking if it's possible for it to be the one that lends the hand and establishes links and bridges.

**Comment by melix** [comment]
I think you're missing the point. Fixing the POM is about fixing when you consume it. Of course it's better to fix it at the producer side, and you should if you can. And of course we acknowledge that it's not always possible. That's why component metadata rules or resolution rules are there: to fix real world problems. It has nothing to do with Gradle metadata as Gradle metadata can have he same issues, and can be fixed the same way, using rules. The point is, one use case = one solution. Sometimes it's going to be "exclude", sometimes not. And our industry will suffer as long as people lazily use "exclude" without thinking about the right solution.

So, your text is actually a great summary of _why_ we do this. This is precisely because there's sh... published everywhere that we provide tools to "fix" metadata. But fixing is not nice. It's better not to make errors in the first place, and that's why we must educate to avoid excludes when you shouldn't.

**Comment by melix** [comment]
As for the actual use cases you listed:

> It should not be there because an alternative exists. The case for commons-logging is clear with slf4j and it's jcl bridge (this is quite an specific case that the logging plugin can solve).

This is a [capability conflict](https://docs.gradle.org/current/userguide/component_capabilities.html). The libraries should declare that they provide the same capability, and the choice should be made by the consumer. Therefore here an _exclude_ is not the answer. The answer is, declare capabilities and make a choice on the consumer side. If the library isn't built with Gradle or is not published with Gradle metadata, component metadata rules can amend existing metadata with the missing information.

> It should not be there because it provides duplicate classes that cause either compile or runtime errors.

There are multiple reasons for this. Either binary incompatibilty, or something else. Depending on the use cases, you may prefer to use [strict dependencies](https://blog.gradle.org/version-downgrade), dependency constraints, ... Incompatibilities are in general poorly modelled via excludes, because it forces you to redeclare dependencies on things you don't directly depend on. It's better to express that "I'm not compatible with X" and let the build tool choose a compatible version.

> It should not be there because it contains classes with unsupported bytecode.

This is a variant of the previous case, which again is fixed by publishing the bytecode level, which is what Gradle metadata does. In case you don't use Gradle metadata, Gradle can patch it for you, as we do for Guava, effectively [fixing the dreaded -jre upgrade issue](https://blog.gradle.org/guava). Said differently, you shouldn't exclude `guava-21-jre` because *you* don't need it. You should say "I need Guava 21" and let the build tool choose between jre and android based on your JVM version (yes, the name Android is misleading in Guava, it actually talks about the bytecode level).

> There may be other reasons.

Yes, and we take care of analyzing each reason and provide the right answer for each of them. There are good reasons to use excludes, as the docs mentioned above explain, but they must be used with care or we're not making progress.

**Comment by vlsi** [comment]
>it knows the resolved artifacts per configuration, this includes direct and transitive dependencies

@aalmiray, there's one more point: Gradle never knows which **repository** will be used for the resolution of the published pom.xml and the transitive dependencies. That is why it can't really predict if the excluded dependency appears or not.

**Comment by aalmiray** [comment]
@vlsi I'm not asking to predict. I'm saying that the all data reachable from the build's POV at some point. Yes, that would require resolving the graph. Yes, potentially a POM file could be served by repository A instead of repository B. Yes, if repository A contains a malicious version of the resolved POM the build will likely make the wrong choice. But if you have malicious dependencies and spurious repositories in your build then you have bigger problems that extra exclusions in your POMs, don't you?

Besides, you _can_ lock dependencies to repositories, this was added some time back in Gradle 5. Which means that if a developer has detected that repository A has served the wrong file he or she can go back and make changes to the build to exclude repository A.

**Comment by ljacomet** [comment]
Closing as there is no plan to have Gradle map configuration level exclude in a more fine grained way.
